### ISSUE TITLE
STDOUT logging on network failures

### ISSUE THREAD

**Issue by itdependsnetworks** [ansible/ansible#26809]
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

network modules

##### ANSIBLE VERSION

```
ansible 2.3.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/etc/ntc/ansible/library', u'/etc/ntc/ntc-ansible-plugins/library']
  python version = 2.7.6 (default, Oct 26 2016, 20:32:47) [GCC 4.8.4]
```

##### SUMMARY

Prior to 2.3, when using network modules (ios, eos, nxos, etc...) that required ssh support, connection failures produced usable output to indicate the connection issue. With the use of persistence that feature has gone away and replaced with a multi-step process to log, then view the logs. While it is understood that you can have access to the logs as described here: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell There is no immediate feedback to the user. 

##### STEPS TO REPRODUCE

Run any playbook with bad credentials or non-existent IP.

```yaml

```

##### EXPECTED RESULTS

Same as 2.2
```
(ansible2.2) ken@kcelenza-ubuntu:/github/cookbook$ ansible-playbook -i inventory ios.yml -vvv
Using /etc/ansible/ansible.cfg as config file

PLAYBOOK: ios.yml **************************************************************
1 plays in ios.yml

PLAY [RUN COMMAND AND SAVE] ****************************************************

TASK [RUN COMMANDS] ************************************************************
task path: /github/cookbook/ios.yml:23
Using module file /home/ken/.virtualenvs/ansible2.2/local/lib/python2.7/site-packages/ansible/modules/core/network/ios/ios_command.py
<10.1.100.46> ESTABLISH LOCAL CONNECTION FOR USER: ken
<10.1.100.46> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1500042841.83-229486189401661 `" && echo ansible-tmp-1500042841.83-229486189401661="` echo ~/.ansible/tmp/ansible-tmp-1500042841.83-229486189401661 `" ) && sleep 0'
Using module file /home/ken/.virtualenvs/ansible2.2/local/lib/python2.7/site-packages/ansible/modules/core/network/ios/ios_command.py
<10.1.100.1> ESTABLISH LOCAL CONNECTION FOR USER: ken
<10.1.100.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1500042841.83-15461642620487 `" && echo ansible-tmp-1500042841.83-15461642620487="` echo ~/.ansible/tmp/ansible-tmp-1500042841.83-15461642620487 `" ) && sleep 0'
<10.1.100.46> PUT /tmp/tmpnNfZ50 TO /home/ken/.ansible/tmp/ansible-tmp-1500042841.83-229486189401661/ios_command.py
<10.1.100.46> EXEC /bin/sh -c 'chmod u+x /home/ken/.ansible/tmp/ansible-tmp-1500042841.83-229486189401661/ /home/ken/.ansible/tmp/ansible-tmp-1500042841.83-229486189401661/ios_command.py && sleep 0'
<10.1.100.1> PUT /tmp/tmpOgUCo4 TO /home/ken/.ansible/tmp/ansible-tmp-1500042841.83-15461642620487/ios_command.py
<10.1.100.1> EXEC /bin/sh -c 'chmod u+x /home/ken/.ansible/tmp/ansible-tmp-1500042841.83-15461642620487/ /home/ken/.ansible/tmp/ansible-tmp-1500042841.83-15461642620487/ios_command.py && sleep 0'
<10.1.100.46> EXEC /bin/sh -c 'python /home/ken/.ansible/tmp/ansible-tmp-1500042841.83-229486189401661/ios_command.py; rm -rf "/home/ken/.ansible/tmp/ansible-tmp-1500042841.83-229486189401661/" > /dev/null 2>&1 && sleep 0'
<10.1.100.1> EXEC /bin/sh -c 'python /home/ken/.ansible/tmp/ansible-tmp-1500042841.83-15461642620487/ios_command.py; rm -rf "/home/ken/.ansible/tmp/ansible-tmp-1500042841.83-15461642620487/" > /dev/null 2>&1 && sleep 0'
fatal: [ios]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "auth_pass": null,
            "authorize": false,
            "commands": [
                "show inventory",
                "show version"
            ],
            "host": "10.1.100.1",
            "interval": 1,
            "match": "all",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": null,
            "provider": {
                "host": "10.1.100.1",
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "username": "ntc"
            },
            "retries": 10,
            "ssh_keyfile": null,
            "timeout": 10,
            "transport": null,
            "use_ssl": true,
            "username": "ntc",
            "validate_certs": true,
            "wait_for": null
        },
        "module_name": "ios_command"
    },
    "msg": "failed to connect to 10.1.100.1:22"
}
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_sPVrBz/ansible_module_ios_command.py", line 237, in <module>
    main()
  File "/tmp/ansible_sPVrBz/ansible_module_ios_command.py", line 200, in main
    runner.add_command(**cmd)
  File "/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/netcli.py", line 147, in add_command
  File "/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/network.py", line 117, in cli
  File "/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/network.py", line 148, in connect
  File "/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/ios.py", line 180, in connect
  File "/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/shell.py", line 226, in connect
  File "/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/shell.py", line 96, in open
  File "/home/ken/.virtualenvs/ansible2.2/local/lib/python2.7/site-packages/paramiko/client.py", line 329, in connect
    raise NoValidConnectionsError(errors)
paramiko.ssh_exception.NoValidConnectionsError: [Errno None] Unable to connect to port 22 on 10.1.100.46

fatal: [ios1]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_name": "ios_command"
    },
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_sPVrBz/ansible_module_ios_command.py\", line 237, in <module>\n    main()\n  File \"/tmp/ansible_sPVrBz/ansible_module_ios_command.py\", line 200, in main\n    runner.add_command(**cmd)\n  File \"/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/netcli.py\", line 147, in add_command\n  File \"/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/network.py\", line 117, in cli\n  File \"/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/network.py\", line 148, in connect\n  File \"/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/ios.py\", line 180, in connect\n  File \"/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/shell.py\", line 226, in connect\n  File \"/tmp/ansible_sPVrBz/ansible_modlib.zip/ansible/module_utils/shell.py\", line 96, in open\n  File \"/home/ken/.virtualenvs/ansible2.2/local/lib/python2.7/site-packages/paramiko/client.py\", line 329, in connect\n    raise NoValidConnectionsError(errors)\nparamiko.ssh_exception.NoValidConnectionsError: [Errno None] Unable to connect to port 22 on 10.1.100.46\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE"
}

PLAY RECAP *********************************************************************
ios                        : ok=0    changed=0    unreachable=0    failed=1
ios1                       : ok=0    changed=0    unreachable=0    failed=1

(ansible2.2) ken@kcelenza-ubuntu:/github/cookbook$

```

##### ACTUAL RESULTS

```
(ansible2.3) ken@kcelenza-ubuntu:/github/cookbook$ ansible-playbook -i inventory ios.yml -vvv
Using /etc/ansible/ansible.cfg as config file

PLAYBOOK: ios.yml **************************************************************************************************************************************************************************************************************************************************************
1 plays in ios.yml

PLAY [RUN COMMAND AND SAVE] ****************************************************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [RUN COMMANDS] ************************************************************************************************************************************************************************************************************************************************************
task path: /github/cookbook/ios.yml:23
fatal: [ios1]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "unable to open shell",
    "rc": 255
}
fatal: [ios]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "unable to open shell",
    "rc": 255
}

PLAY RECAP *********************************************************************************************************************************************************************************************************************************************************************
ios                        : ok=0    changed=0    unreachable=0    failed=1
ios1                       : ok=0    changed=0    unreachable=0    failed=1

(ansible2.3) ken@kcelenza-ubuntu:/github/cookbook$ ansible --version
ansible 2.3.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/etc/ntc/ansible/library', u'/etc/ntc/ntc-ansible-plugins/library']
  python version = 2.7.6 (default, Oct 26 2016, 20:32:47) [GCC 4.8.4]
(ansible2.3) ken@kcelenza-ubuntu:/github/cookbook$ workon ansible2.4
(ansible2.4) ken@kcelenza-ubuntu:/github/cookbook$ ansible-playbook -i inventory ios.yml -vvv
ansible-playbook 2.4.0 (kc_pan_sec_rule_idemp 7ee2685edb) last updated 2017/07/09 23:25:10 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/etc/ntc/ansible/library', u'/etc/ntc/ntc-ansible-plugins/library']
  ansible python module location = /home/ken/ansible/lib/ansible
  executable location = /home/ken/.virtualenvs/ansible2.4/bin/ansible-playbook
  python version = 2.7.6 (default, Oct 26 2016, 20:32:47) [GCC 4.8.4]
Using /etc/ansible/ansible.cfg as config file
Parsed /github/cookbook/inventory inventory source with ini plugin

PLAYBOOK: ios.yml **************************************************************************************************************************************************************************************************************************************************************
1 plays in ios.yml

PLAY [RUN COMMAND AND SAVE] ****************************************************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [RUN COMMANDS] ************************************************************************************************************************************************************************************************************************************************************
task path: /github/cookbook/ios.yml:23
<10.1.100.1> using connection plugin network_cli
<10.1.100.46> using connection plugin network_cli
fatal: [ios1]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"
}
fatal: [ios]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"
}

PLAY RECAP *********************************************************************************************************************************************************************************************************************************************************************
ios                        : ok=0    changed=0    unreachable=0    failed=1
ios1                       : ok=0    changed=0    unreachable=0    failed=1

```

pinging @privateip @gundalow @rcarrillocruz

**Comment by wolfman2g1** [comment]
+1

**Comment by itdependsnetworks** [comment]
The link for this is broken for the last (    https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell ) day. Which highlights the reason this should be a priority.

**Comment by ktbyers** [comment]
This is a really big issue for newcomers. 

It was my understanding from previous communication that this was going to be greatly improved in Ansible 2.4, but from what I can see it is the same.

Also, the logs can be close to unintelligible. For example, a very basic failure (i.e. a missing entry in SSH known_hosts results in this). 

```
2017-08-28 11:42:38,412 p=1878 u=kbyers |  connecting to host cisco1.domain.com returned an error
2017-08-28 11:42:38,413 p=1878 u=kbyers |  (14, 'Bad address')
```

This is on Ansible 2.4 devel branch and on Ansible 2.3.1.

**Comment by gundalow** [comment]
Update
This **is** something we are fixing for `2.5.0`, the initial work for this is https://github.com/ansible/ansible/issues/31107 (implemented in https://github.com/ansible/ansible/pull/31215)

Once this is merged into `devel` we welcome your feedback on the way error messages are reported

**Comment by privateip** [comment]
This work was completed as part of ##32521
